# Assignment 2

## Instructions - Read this first!

This is an individual homework assignment. This means that:

- You may discuss the problems in this assignment with other students in this course and your instructor/TA, but YOUR WORK MUST BE YOUR OWN.
- Do not show other students code or your own work on this assignment.
- You may consult external references, but not actively receive help from individuals not involved in this course.
- Cite all references outside of the course you used, including conversations with other students which were helpful. (This helps us give credit where it is due!). All references must use a commonly accepted reference format, for example, APA or IEEE (or another citation style of your choice).

If any of these rules seem ambiguous, please check with with your instructor for help interpreting them.

We suggest completing this assignment using the provided notebook. Each question should be answered using a SQL query (or combination or SQL queries) unless the text indicates that you may do something else. You may submit your queries embedded in Python, using SQLAlchemy or the MySQL Connector, or as plain text in Markdown.

## When you submit your work

Your submission will be graded manually. To ensure that everything goes smoothly, please follow these instructions to prepare your notebook for submission to the D2L Dropbox for Assignment 2:

- Please remove any print statments used to test your work (you can comment them out)
- Please provide your solutions where asked; please do not alter any other parts of this notebook.
- If you need to add cells to test your code please move them to the end of the notebook before submission- or you may included your commented out answers and tests in the cells provided

## Introduction

 In this assignment, we will focus familiarizing you with using SQL for data exploration, and continuing to cultivate a sense of curiosity about the datasets you encounter. We will be using a table generated by Statistics Canada (2022), based upon census data, about the knowledge of languages by area in Canada. This table has been pre-cleaned (although potentially not entirely) so that you can work on the actual assignment more quickly.
 
 To begin, start by importing the provided CSV into your own SQL database using SQLAlchemy, by filling in the lines below:

In [2]:
import pandas as pd
import sqlalchemy as sq


# read in your CSV as a dataframe
language = pd.read_csv("language_speakers.csv")

# connect to your database; include a cell at the bottom of this notebook to dispose of your engine object
engine = sq.create_engine('mysql+mysqlconnector://kane_smith1:8ULM75GHS@datasciencedb2.ucalgary.ca/kane_smith1')

# write your dataframe into a table
#language.to_sql('language_locations', engine )
language_df = pd.read_sql_table("language_locations", engine)
language_df.head()
# demonstrate that your import has been successful by reading your database table as a dataframe, 
# and print some information (not the entire table) about your second dataframe

index  REF_DATE     GEO           DGUID Statistics (3)  \
0      0      2021  Canada  2021A000011124          Count   
1      1      2021  Canada  2021A000011124          Count   
2      2      2021  Canada  2021A000011124          Count   
3      3      2021  Canada  2021A000011124          Count   
4      4      2021  Canada  2021A000011124          Count   

     Knowledge of languages (324)  \
0  Total - Knowledge of languages   
1  Total - Knowledge of languages   
2  Total - Knowledge of languages   
3              Official languages   
4              Official languages   

  Single and multiple responses of knowledge of languages (3) SCALAR_FACTOR  \
0  Total - Single and multiple responses of knowl...                  units   
1         Single responses of knowledge of languages                  units   
2       Multiple responses of knowledge of languages                  units   
3  Total - Single and multiple responses of knowl...                  units   
4         Single responses of knowledge of languages                  units   

   SCALAR_ID   COORDINATE     VALUE  
0          0  1.1.1.1.1.1  36328480  
1          0  1.1.1.1.1.2  21377910  
2          0  1.1.1.1.1.3  14950565  
3          0  1.1.1.1.2.1  35644655  
4          0  1.1.1.1.2.2  20797495

## Part A: Warm-up Questions (12 marks)

Answer the questions below, including the queries you used where necessary. Not all questions will require writing a SQL query to answer.

**(1 mark)**

How many records are there in total?

In [54]:
query_table1 = pd.read_sql_query('''
SELECT COUNT(*) AS NUM_RECORDS
FROM language_locations;''', engine)
display(query_table1)

NUM_RECORDS
0         7536

**(1 mark)**

How many different languages are spoken in Canada?

In [56]:
'''
Explanation:
There are some rows in the data that are not actual languages, but are totals of sub-categories of languages. To get the number of unique languages spoken, these rows should be removed.
eg. "Official Languages" is not an actual language, but a total of the "English" and "French" languages.
'''
query_table2 = pd.read_sql_query('''
SELECT GEO AS LOCATION, COUNT(DISTINCT `Knowledge of languages (324)`) AS NUM_DISTINCT_LANGUAGES
FROM language_locations
WHERE `Knowledge of languages (324)` NOT LIKE '%languages';''', engine)
display(query_table2)

LOCATION  NUM_DISTINCT_LANGUAGES
0   Canada                     263

**(1 mark)**

Is there data for all languages, for each province or territory?

In [61]:
'''
Explanation:
To get the number of unique languages spoken for each Province/Territory, we can group by GEO. We should exclude "Canada" as it is not a Province/Territory, but rather a 
total of all the Province/Territories. 
'''
query_table3 = pd.read_sql_query('''
SELECT GEO AS PROVINCE, COUNT(DISTINCT `Knowledge of languages (324)`) AS COUNT_DISTINCT_LANGUAGES
FROM language_locations
WHERE `Knowledge of languages (324)` NOT LIKE '%languages' AND GEO NOT LIKE 'Canada'
GROUP BY GEO;''', engine)
display(query_table3)

PROVINCE  COUNT_DISTINCT_LANGUAGES
0                     Alberta                       235
1            British Columbia                       252
2                    Manitoba                       203
3               New Brunswick                       156
4   Newfoundland and Labrador                       124
5       Northwest Territories                       102
6                 Nova Scotia                       169
7                     Nunavut                        62
8                     Ontario                       234
9        Prince Edward Island                       100
10                     Quebec                       220
11               Saskatchewan                       197
12                      Yukon                        94

Since each Province/Territory does not have 263 (found in previous question) unique languages spoken, there is <u>not</u> data for all languages for each Province/Territory. 

**(8 marks)**

Explain what each of the following columns is used for. You may use the original page to guide your explanation (but you should cite it)
* GEO:
* Statistics:
* Knowledge of languages:
* Single and multiple responses of knowledge of languages (2 marks):
* COORDINATE (2 marks):
* VALUE: 

* **GEO:** (string) Has the province and territory data
* **Statistics:** (string) The type of statistic being used in single response, multiple response, and total of knowledge of languages.  
* **Knowledge of languages:** (string) The identifier of language responded to in the census.
* **Single and multiple responses of knowledge of languages (2 marks):** (string) The identifier of where the value is for single or multiple response of knowledge of languages. There is also a Total, which is the sum of the Single and Multiple language speakers. So there are 3 rows for each language for each Province/Territory: Single language speakers who speak the language, multiple language speakers who speak the language and the total.
* **COORDINATE (2 marks):** (string) This looks to be the unique key in the table. It uniquely identifies each observation in the dataset. You can also identify the Province/Territory, Single, Multiple, or Total language speakers and the language being spoken based on certain parts of the coordinate value. 
* **VALUE:** (numeric) The value of the statistic (count) stated in the "Statistics" column 

**(1 mark)**

Is it possible to tell how many individuals live in Canada based on this dataset? Why or why not?

Yes, <u>assuming the census includes responses from all individuals in Canada</u>. The count of **"Total - Knowledge of Languages"** and **"Total - Single and Multiple responses of languages"** will provide the total count of respondants to the census. **(Which is 36,328,480).**

## Part B: Simple questions (10 marks) 

For these queries, run a query which provides the answer.

**(2 marks)**

Which province or territory has the highest number of individuals who can speak multiple languages?

In [59]:
'''
Explanation:
To get the number of people who speak multiple languages for each Province/Territory, I filtered rows to be only multiple language speakers and the "Knowledge of languages" 
for the column "Knowledge of languages (324)". I then grouped by Province/Territory, excluding "Canada" and then found the maximum value for number of multiple language speakers.
'''
query_table4 = pd.read_sql_query('''
SELECT GEO AS PROVINCE, MAX(VALUE) AS MAX_NUM_MULTI_LANG_SPEAKERS
FROM language_locations
WHERE `Single and multiple responses of knowledge of languages (3)` LIKE 'Multiple responses of knowledge of languages' AND GEO NOT LIKE 'Canada' AND  `Knowledge of languages (324)` LIKE 'Total - Knowledge of languages'
GROUP BY GEO
ORDER BY MAX(VALUE) DESC;''', engine)
display(query_table4)

PROVINCE  MAX_NUM_MULTI_LANG_SPEAKERS
0                     Ontario                      5855325
1                      Quebec                      4564580
2            British Columbia                      1870730
3                     Alberta                      1354235
4                    Manitoba                       458075
5               New Brunswick                       296330
6                Saskatchewan                       245430
7                 Nova Scotia                       178055
8   Newfoundland and Labrador                        43800
9        Prince Edward Island                        32980
10                    Nunavut                        25715
11      Northwest Territories                        13510
12                      Yukon                        11815

<u>**Ontario**</u> has the highest number of indivudals who can speak at least 2 languages.

**(2 marks)**

Which provinces or territories has the most diverse number of languages spoken?

In [60]:
'''
Explanation:
To find the Province with the highest amount of unique language spoken, I used COUNT(DISINCT) of the `Knowledge of languages (324)` while excluding anything like '%languages in it, as 
these are totals of languages instead of actual languages. I then grouped by Province/Territory and ordered the count in descending order so the highest would be at the top.
'''
query_table5 = pd.read_sql_query('''
SELECT GEO AS PROVINCE, COUNT(DISTINCT `Knowledge of languages (324)`) AS COUNT_DISTINCT_LANG
FROM language_locations
WHERE `Knowledge of languages (324)` NOT LIKE '%languages' AND GEO NOT LIKE 'Canada'
GROUP BY GEO
ORDER BY COUNT_DISTINCT_LANG DESC;''', engine)
display(query_table5)

PROVINCE  COUNT_DISTINCT_LANG
0            British Columbia                  252
1                     Alberta                  235
2                     Ontario                  234
3                      Quebec                  220
4                    Manitoba                  203
5                Saskatchewan                  197
6                 Nova Scotia                  169
7               New Brunswick                  156
8   Newfoundland and Labrador                  124
9       Northwest Territories                  102
10       Prince Edward Island                  100
11                      Yukon                   94
12                    Nunavut                   62

<u>**British Columbia**</u> has the highest number of unique languages spoken by individuals

**(2 marks)**

Which provinces or territories have the most diverse number of Indigenous languages spoken?

In [3]:
'''
Explanation:
Indigenous languages has a coordinate containing .7. to .96. (eg. Indigenous languages in Canada have coordinates from 1.1.1.1.7 to 1.1.1.1.93. 
We can parse the coordinate column to get just the last part of the coordinate to indentify the indigenous languages for each Province/Territory.
'''
query_table6 = pd.read_sql_query('''
SELECT GEO AS PROVINCE, COUNT(DISTINCT `Knowledge of languages (324)`) AS COUNT_DISTINCT_LANG
FROM language_locations
WHERE  `Knowledge of languages (324)` NOT LIKE '%languages' AND GEO NOT LIKE 'Canada' AND SUBSTRING_INDEX(SUBSTRING_INDEX(COORDINATE, ".", -2), ".", 1) BETWEEN 7 AND 96
GROUP BY GEO
ORDER BY COUNT_DISTINCT_LANG DESC ;''', engine)
display(query_table6)

PROVINCE  COUNT_DISTINCT_LANG
0            British Columbia                   63
1                     Alberta                   47
2                     Ontario                   44
3                      Quebec                   33
4                    Manitoba                   21
5                Saskatchewan                   21
6       Northwest Territories                   19
7                       Yukon                   17
8                 Nova Scotia                   13
9               New Brunswick                   11
10  Newfoundland and Labrador                    9
11                    Nunavut                    7
12       Prince Edward Island                    2

<u>**British Columbia**</u> has the most diverse number of indigenous languages spoken with 63.

**(2 marks)**

As a percentage of all language speakers in each province or territory, how commonly is French spoken?

In [63]:
''' 
Explanation: 
To answer this, I will take the total amount of French speakers (single and multiple knowedgle of languages) and divide it by 
Total-Knowledge of languages (single and multiple knowedgle of languages) for each Province/Territory.
'''
query_table7 = pd.read_sql_query('''
SELECT  t1.GEO AS PROVINCE, (t2.TOTAL/t1.TOTAL)*100 AS PROP_FRENCH
FROM 
(SELECT GEO, VALUE AS TOTAL
FROM language_locations
WHERE `Knowledge of languages (324)` LIKE 'French' AND GEO NOT LIKE 'Canada' AND `Knowledge of languages (324)` NOT LIKE '%languages' AND `Single and multiple responses of knowledge of languages (3)` LIKE 'Total - Single and multiple responses of knowledge of languages'
GROUP BY GEO) AS t2

JOIN (SELECT GEO, VALUE AS TOTAL
FROM language_locations
WHERE GEO NOT LIKE 'Canada' AND `Knowledge of languages (324)` LIKE 'Total - Knowledge of languages' AND `Single and multiple responses of knowledge of languages (3)` LIKE 'Total - Single and multiple responses of knowledge of languages'
GROUP BY GEO) AS t1

ON t1.GEO = t2.GEO
ORDER BY PROP_FRENCH DESC;''', engine)
display(query_table7)

PROVINCE  PROP_FRENCH
0                      Quebec        93.72
1               New Brunswick        41.86
2                       Yukon        14.50
3        Prince Edward Island        12.92
4                     Ontario        11.05
5       Northwest Territories        11.01
6                 Nova Scotia        10.39
7                    Manitoba         8.55
8            British Columbia         6.66
9                     Alberta         6.23
10  Newfoundland and Labrador         5.20
11               Saskatchewan         4.72
12                    Nunavut         3.97

<u>**Quebec**</u> has the highest proportion of French speakers at 93.72%

**(2 marks)**

As a percentage of all Chinese speakers in the country, how many Chinese speakers do not speak Mandarin?

In [65]:
''' 
Explanation: 
To answer this, I found the total number of Mandarin speakers (single and multiple language) and the total number of Chinese language speakers(single and multiple language)
then I used the formula [(Chinese language speakers - Mandarin speakers) / (Chinese language speakers)] to get the proportion of Chinese speakers who do not speak Mandarin.
'''
query_table8 = pd.read_sql_query('''
SELECT  t1.GEO AS PROVINCE, ((t2.TOTAL-t1.TOTAL)/t2.TOTAL)*100 AS PROP_NOT_MANDARIN
FROM 
(SELECT GEO, VALUE AS TOTAL
FROM language_locations
WHERE `Knowledge of languages (324)` LIKE 'Mandarin' AND GEO LIKE 'Canada' AND `Single and multiple responses of knowledge of languages (3)` LIKE 'Total - Single and multiple responses of knowledge of languages') AS t1

JOIN (SELECT GEO, SUM(VALUE) AS TOTAL
FROM language_locations
WHERE GEO LIKE 'Canada' AND `Knowledge of languages (324)` LIKE 'Chinese languages' AND `Single and multiple responses of knowledge of languages (3)` LIKE 'Total - Single and multiple responses of knowledge of languages') AS t2

ON t1.GEO = t2.GEO;''', engine)
display(query_table8)

PROVINCE  PROP_NOT_MANDARIN
0   Canada              35.42

The percentage of Chinese speakers who do <u> not</u> speak Mandarin in Canada is 35.42%

## Part C: Detailed analysis (20 marks)

Now consider being given a task to make sense of the distribution of languages spoken in Canada using this dataset.

**Question 1 (4 marks)**

Create two guiding questions to use in your analysis, and include them below as Markdown. As a starting point (and remember you are not limited to only these!), you may want to consider the following ideas:
- focus on a specific set of languages which are interesting to you
- consider the effect of language speakers who speak only one or multiple languages
- differences between specific provinces and territories in terms of patterns of language speakers

**Guiding Question 1:** Ukrainian is considered an Indo-European language. What is the prevelance of Ukrainian across each Province/Territory and what is the top Indo-European language spoken for each Province/Territory?


**Guiding Question 2:** What are the top 5 languages that have the highest prevelance of single language speakers? What Province/Territory is the top language most spoken by  the population?

**Question 2 (12 marks)** 

Write at least four queries (two for each question) which you believe will address one of your guiding questions. Clearly indicate which queries address your questions. You may wish to include a comment to explain why this query will help address your question.

In [5]:
# Guiding Question 1 Queries
query_table9 = pd.read_sql_query('''
SELECT  t1.GEO AS PRVOINCE, (t2.TOTAL/t1.TOTAL)*100 AS PROP_UKRAINIAN
FROM 
(SELECT GEO, VALUE AS TOTAL
FROM language_locations
WHERE `Knowledge of languages (324)` LIKE 'Ukrainian' AND GEO NOT LIKE 'Canada' AND `Knowledge of languages (324)` NOT LIKE '%languages' AND `Single and multiple responses of knowledge of languages (3)` LIKE 'Total - Single and multiple responses of knowledge of languages'
GROUP BY GEO) AS t2

JOIN (SELECT GEO, VALUE AS TOTAL
FROM language_locations
WHERE GEO NOT LIKE 'Canada' AND `Knowledge of languages (324)` LIKE 'Total - Knowledge of languages' AND `Single and multiple responses of knowledge of languages (3)` LIKE 'Total - Single and multiple responses of knowledge of languages'
GROUP BY GEO) AS t1

ON t1.GEO = t2.GEO
ORDER BY PROP_UKRAINIAN DESC;''', engine)
display(query_table9)

print('''
Explanation:
This query gives the proportion of Indo-European languages that is Ukranian for each Province/Territory and sorts
it from greatest to least. This allows me to answer my guiding question of which Province/Territory has the 
highest proportion of Indo-European languages that is Ukranian.
''')


query_table10 = pd.read_sql_query('''
SELECT GEO AS PROVINCE,`Knowledge of languages (324)` AS LANGUAGE
FROM language_locations
WHERE GEO NOT LIKE 'Canada' AND SUBSTRING_INDEX(SUBSTRING_INDEX(COORDINATE, ".", -2), ".", 1) BETWEEN 165 AND 186 AND `Single and multiple responses of knowledge of languages (3)` LIKE 'Total - Single and multiple responses of knowledge of languages' AND `Knowledge of languages (324)` NOT LIKE '%languages'
GROUP BY GEO;''', engine)
display(query_table10)

print('''
Explanation:
This query returns the most popular Indo-European language for each Province/Territory. This allows me to answer 
second part of my first guiding question.
''')


PRVOINCE  PROP_UKRAINIAN
0                    Manitoba            1.14
1                Saskatchewan            0.98
2                     Alberta            0.56
3                     Ontario            0.42
4            British Columbia            0.25
5                       Yukon            0.18
6               New Brunswick            0.13
7                      Quebec            0.12
8       Northwest Territories            0.06
9                 Nova Scotia            0.06
10       Prince Edward Island            0.06
11                    Nunavut            0.04
12  Newfoundland and Labrador            0.02


Explanation:
This query gives the proportion of Indo-European languages that is Ukranian for each Province/Territory and sorts
it from greatest to least. This allows me to answer my guiding question of which Province/Territory has the 
highest proportion of Indo-European languages that is Ukranian.



PROVINCE  LANGUAGE
0                     Alberta  Albanian
1            British Columbia  Albanian
2                    Manitoba  Albanian
3               New Brunswick  Albanian
4   Newfoundland and Labrador  Albanian
5       Northwest Territories  Armenian
6                 Nova Scotia  Albanian
7                     Nunavut    Polish
8                     Ontario  Albanian
9        Prince Edward Island  Albanian
10                     Quebec  Albanian
11               Saskatchewan  Albanian
12                      Yukon     Czech


Explanation:
This query returns the most popular Indo-European language for each Province/Territory. This allows me to answer 
second part of my first guiding question.



**<u>Analysis:</u>**

**Manitoba** has the <u> highest </u> proportion of Ukrainian speakers at 1.14%, while **Newfoundland and Labrador** has the <u>lowest</u> proportion of Ukrainian speakers at 0.02%.

For most of the Provinces/Territories, Albanian is the most common Indo-European language spoken. With the following exceptions:
- Armenian is the most common in  Northwest Territories
- Czech is the most common in the Yukon
 - Polish is the most common in Nunavut

In [71]:
# Guiding Question 2 Queries 
query_table11 = pd.read_sql_query('''
SELECT t1.LANG AS LANGUAGE, (t2.TOTAL/t1.TOTAL)*100 AS PROP_SINGLE_LANG
FROM 
(SELECT GEO, SUM(VALUE) AS TOTAL, `Knowledge of languages (324)` AS LANG
FROM language_locations
WHERE `Knowledge of languages (324)` NOT LIKE '%languages' AND GEO NOT LIKE 'Canada' AND `Single and multiple responses of knowledge of languages (3)` LIKE 'Single responses of knowledge of languages'
GROUP BY LANG) AS t2

INNER JOIN (SELECT GEO, SUM(VALUE) AS TOTAL, `Knowledge of languages (324)` AS LANG
FROM language_locations
WHERE GEO NOT LIKE 'Canada' AND `Knowledge of languages (324)` NOT LIKE '%languages' AND `Single and multiple responses of knowledge of languages (3)` LIKE 'Total - Single and multiple responses of knowledge of languages'
GROUP BY LANG) AS t1

ON t2.LANG = t1.LANG
ORDER BY PROP_SINGLE_LANG DESC''', engine)
display(query_table11)

print('''
Explanation:
This query shows me the proportion of single-language speakers for each language spoken in Canada and sorts it from
greatest to least. This allows me to answer the part of my guiding question of which language has the highest
proportion of single-language speakers.
''')

query_table12 = pd.read_sql_query('''
SELECT  t1.GEO AS PROVINCE, (t2.TOTAL/t1.TOTAL)*100 AS PROP_ENGLISH
FROM 
(SELECT GEO, VALUE AS TOTAL
FROM language_locations
WHERE `Knowledge of languages (324)` LIKE 'English' AND GEO NOT LIKE 'Canada' AND `Knowledge of languages (324)` NOT LIKE '%languages' AND `Single and multiple responses of knowledge of languages (3)` LIKE 'Total - Single and multiple responses of knowledge of languages'
GROUP BY GEO) AS t2

JOIN (SELECT GEO, VALUE AS TOTAL
FROM language_locations
WHERE GEO NOT LIKE 'Canada' AND `Knowledge of languages (324)` LIKE 'Total - Knowledge of languages' AND `Single and multiple responses of knowledge of languages (3)` LIKE 'Total - Single and multiple responses of knowledge of languages' 
GROUP BY GEO) AS t1

ON t1.GEO = t2.GEO
ORDER BY PROP_ENGLISH DESC''', engine)
display(query_table12)

print('''
Explanation:
This query gives the proportion of the population that speaks English for each Province/Territory and sorted from 
greatest to least. This allows me to answer the second part of my second guiding question of which Province/Territory 
the highest proportion of of English speakers in their population. 
''')

LANGUAGE  PROP_SINGLE_LANG
0                              English             54.44
1                               French             33.88
2               Sign languages, n.i.e.             21.22
3                             Rohingya             20.74
4                          S'gaw Karen             14.61
..                                 ...               ...
171                            Konkani              0.17
172                              Malay              0.14
173                         Hiligaynon              0.12
174  Pampangan (Kapampangan, Pampango)              0.09
175                               Igbo              0.08

[176 rows x 2 columns]


Explanation:
This query shows me the proportion of single-language speakers for each language spoken in Canada and sorts it from
greatest to least. This allows me to answer the part of my guiding question of which language has the highest
proportion of single-language speakers.



PROVINCE  PROP_ENGLISH
0   Newfoundland and Labrador         99.81
1                 Nova Scotia         99.59
2        Prince Edward Island         99.37
3                       Yukon         99.37
4                Saskatchewan         99.24
5       Northwest Territories         99.05
6                    Manitoba         98.60
7                     Alberta         98.37
8                     Ontario         97.28
9            British Columbia         96.69
10                    Nunavut         94.14
11              New Brunswick         91.94
12                     Quebec         51.96


Explanation:
This query gives the proportion of the population that speaks English for each Province/Territory and sorted from 
greatest to least. This allows me to answer the second part of my second guiding question of which Province/Territory 
the highest proportion of of English speakers in their population.



**<u>Analysis:</u>**

The language with the highest proportion of speaker who only speak that language is **English** at 54.44%. 

The Province/Territory with the highest proportion of English speaker is **Newfoundland and Labrador** at 99.81%. Interestingly enough, all Provinces/Territories have over a 90% English speaking population except for Quebec, which is almost at 50%.|

**Question 3 (4 marks)**

What kind of data would be interesting to receive in combination with the data in this table? Discuss how you could use this additional information to extend one of your guiding questions.

**Answer:**
I think having immigration data for each Province/Territory would be interesting to combine with this language data. For my first guiding question, I would be able to see if the proportion of Ukrainian speakers is correlated with Ukranian immigration in each Province/Territory. Or I could see if the most spoken Indo-European language for each Province/Territory is the same as the most immigrantion to that Province/Territory from the  language-speaking region of the world.

## Part D: Reflection (5 marks)

In 100 to 250 words, identify a concept you have found difficult or confusing from this assignment. Reflect on how your previous learning or experience helped you to understand this concept. Provide your reflection using markdown in the cell below

**Reflection:**
The most difficult part for me was identifying the Indigenous languages in the data. The only thing I found that identifies the Indigenous languages was the pattern of the COORDINATE column. Noticing this, I was able to parse the COORDINATE string to get the meaningful data and filter the data to only have Indigenous languages. I was able to use past knowledge on string manipulation in SQL from my undergraduate degree to help me identify this and put my idea into practice.

In [ ]:
# Use this cell to include some code to dispose of your SQLAlchemy engine object
engine.dispose()

## References

Statistics Canada (Government of Canada). 2022. Knowledge of languages by age and gender: Canada, provinces and territories, census metropolitan areas and census agglomerations with parts. Retrieved October 15, 2022 from https://www150.statcan.gc.ca/t1/tbl1/en/tv.action?pid=9810021701